### ISSUE TITLE
Pull Requests should always be updated such that they preserve history

### ISSUE THREAD

**Issue by sstamenova** [onnx/onnx-mlir#1031]
Recently I've noticed people are updating their PRs by force pushing in such a way that the previous commits disappear. This means a reviewer doesn't have an easy way to compare what was changed between the two versions of the PR making it really difficult to track updates. In general, I would expect that PRs should always preserve history so that changes can be compared easily and things don't get missed. Is there a different paradigm expected for onnx-mlir now?

**Comment by gongsu832** [comment]
This is likely because starting recently (about two or three weeks ago) there seemed to be some problem with the PRs not keeping track of changes in the target master (now main) branch.

For example, I submitted a PR1 at some point against main. Then Kevin submitted another PR2 afterwards and PR2 got merged into main first. So I rebased my PR1 to the new main which now includes commits from PR2. However, my PR1 doesn't recognize that PR2 is now part of the main. Instead, it thinks all the commits of PR2 are part of my PR1. That completely screws up the history of my PR1. So I had to force push to get rid of the commits of PR2 to clean things up.

There are also times where git simply screwed up the commit history of a PR after a lot of merging (of other PRs) and rebasing. So I don't think people are intentionally using forced push. Most likely it's due to git somehow screwing up the PR commit history.

**Comment by etiotto** [comment]
My issue with merging rather than rebasing is that the PR will should several partial commits that are meaningless in the history, in addition it will also pollute the history with spurious informations from merging base. I prefer to rebase all my PRs.

**Comment by AlexandreEichenberger** [comment]
I know that there is a recommendation to use rebase, but it seems to me that the "merge" approach works well because when we finally push the changes of a PR to the main branch, we use a "squash and merge" that appears to put all the PR changes after any of the current one on main.

I feel that removing the superfluous history commit (e.g. "I am fixing for the nth time a formatting issue") could be removed using the `rebase -i`. I can try it on my next PR when only merge are used in the history. Think it works, better to double check.

I am not a sophisticated git user, I am happy to follow the recommendations of experts.

**Comment by AlexandreEichenberger** [comment]
This PR is a continuation of #955

**Comment by etiotto** [comment]
@AlexandreEichenberger looking at recent commits I see that a "squashed and merge" commit has a commit msg that contains the list of msgs that belonged to the individual commits before they were squashed, example (not pointing fingers - just took a random recent commit as example here): 
```
commit 0d483afdcd4696c7d14c562bbf02c170b802bb29
Author: chentong319 <chentong@us.ibm.com>
Date:   Fri Dec 10 21:38:47 2021 -0500

    Instrumentation into blocks (#1055)
    
    * walk
    
    Signed-off-by: Tong Chen <chentong@us.ibm.com>
    
    * fix
    
    Signed-off-by: Tong Chen <chentong@us.ibm.com>
    
    * format
    
    Signed-off-by: Tong Chen <chentong@us.ibm.com>
    
    Co-authored-by: Alexandre Eichenberger <alexe@us.ibm.com>
```

I think the commit should only contain a single message that summarizes the change done. Like: 

```
commit 6f55f8ec7b1cea3eb6d4b714150966c66fb93645
Author: gongsu832 <gong_su@hotmail.com>
Date:   Wed Dec 8 15:17:38 2021 -0500

    Add onnx definitions and includes to PyRuntime (#1047)
    
    Signed-off-by: Gong Su <gong_su@hotmail.com>
    
    Co-authored-by: Alexandre Eichenberger <alexe@us.ibm.com>
```

I am not sure how github decide to merge the various commits, in particular how it decides to form the message associated with the commit (my guess is that it just collates all the individual commit msgs). 

The approach I described here https://github.com/onnx/onnx-mlir/issues/1031#issuecomment-992621526  bypasses completely the issue because it is the author that would grantee the single commit would have an appropriate, concise msg.

**Comment by sstamenova** [comment]
There are two things to consider here:
1) The history of the main branch *after* commit - this is what @etiotto is trying to guard
2) The readability of the PR itself *while* the PR is happening - this is what I would like to see improve

In a perfect world, everyone's commits to their PR branch will be clean and descriptive, so when github squashes them all into a single one, the comments will make sense, however, we don't live in a perfect world and sometimes those commits are small or have bad messages, etc. So it makes sense to squash everything together at the end but before completing the pull request in cases where the history is messy. This will achieve the desired goal of keeping a clean history while also giving code reviewers the ability to see the differences at each step of the PR.

I would like to emphasize how important it is for reviewers to be able to see what has changed in a PR between each commit. Otherwise, each reviewer really should look through the code every time as if it was brand new code which means things they had already noticed but haven't been addressed can get missed as code moves between files, or functions, etc.. The more PRs there are, the more important it is that each PR has a history that can be reviewed.

**Comment by etiotto** [comment]
Hey @sstamenova, I see now what you are suggesting. I agree that is useful to reviewers to see the history of changes during code review.  We could attempt to do the following:
  - create new PRs as drafts, allow rebasing commits until the PR is ready for code review
  - once the PR is ready for code review, change it to a regular PR. Reviewers are free to add comments at this point
  - when the PR is in review refrain from squashing commits, this should preserve history while reviews are in progress
  - when the PR has been approved, allow authors to again squash commits and edit the commit msgs
  - in its final approved for the PR should contain a single commit, with a single meaningful message
  
I think this would achieve the 2 goals we have (keep the history clean and allow reviewers to see changes during review. 
What do you think?

**Comment by sstamenova** [comment]
That process makes sense.

**Comment by etiotto** [comment]
@AlexandreEichenberger seems like we have agreement. What are the next steps. Do we have somewhere we can document the stuff above (assuming you agree as well of course).

**Comment by chentong319** [comment]
From the discussion, I can see the benefit of using rebase. Let's update the document.

**Comment by AlexandreEichenberger** [comment]
Make sense to me. Please write a nice step by step using git commands (and/or which GitHub operations are allowed/should not be allowed). Natural place would be on the doc, under contributing.

Thanks to all that contributed to finding a compromise that work for all.

**Comment by AlexandreEichenberger** [comment]
By the way, I found an incredible easy way to achieve the stated goals from the GitHub interface.

Only do git merge during development and review... full history is preserved.
Hit "merge" button from interface as many times as needed.

When you are ready to squash and merge, the interface actually list in a text box the actual text that will be put in the log. By default, it is the current list of commit, but one can edit it to whatever one wants. So remove all but the first author/date and last signed-off by... and write your harts content in the  middle.

```
Author: Alexandre Eichenberger <alexe@us.ibm.com>
Date:   Wed Jan 19 15:33:14 2022 +0000

<< edit here whatever you want>>
    
    Signed-off-by: Alexandre Eichenberger <alexe@us.ibm.com>
```

Thus, we have the nice text, and the "squash" of the "squash and merge" button already present the changes as only one "event" in the history.

So no need to do any git commit, merge, rebase, if one is content to use the GitHub interface to finalize the text on Github

**Comment by etiotto** [comment]
I think we can close this discussion. I used the what @AlexandreEichenberger suggests and works great. So we have a way to merge with a clean log.  Now we just need to do it consistently.
